packages/remix-dev/vite/plugin.ts

```diff
@@ -1483,46 +1501,72 @@ async function getRouteMetadata(
 }
 
 async function handleSpaMode(
+  manifest: Manifest,
+  future: ResolvedRemixVitePluginConfig["future"],
   serverBuildDirectoryPath: string,
   serverBuildFile: string,
   assetsBuildDirectory: string,
-  viteConfig: Vite.ResolvedConfig
+  viteConfig: Vite.ResolvedConfig,
+  libraryName: string | undefined
 ) {
-  // Create a handler and call it for the `/` path - rendering down to the
-  // proper HydrateFallback ... or not!  Maybe they have a static landing page
-  // generated from routes/_index.tsx.
-  let serverBuildPath = path.join(serverBuildDirectoryPath, serverBuildFile);
-  let build = await import(url.pathToFileURL(serverBuildPath).toString());
-  let { createRequestHandler: createHandler } = await import("@remix-run/node");
-  let handler = createHandler(build, viteConfig.mode);
-  let response = await handler(new Request("http://localhost/"));
-  let html = await response.text();
-  if (response.status !== 200) {
-    throw new Error(
-      `SPA Mode: Received a ${response.status} status code from ` +
-        `\`entry.server.tsx\` while generating the \`index.html\` file.\n${html}`
+  if (libraryName) {
+    let jsDir = path.join(assetsBuildDirectory, "assets");
+    let jsPath = path.join(jsDir, libraryName);
+    let js = remixEntryJsString(
+      // @ts-expect-error
+      manifest,
+      ["root"],
+      createServerHandoffString({
+        state: {},
+        url: "http://localhost/",
+        future,
+        isSpaMode: true,
+      })
     );
-  }
+    await fse.writeFile(jsPath, js);
```

Comment:
Best viewed with whitespace changed hidden, but when `library` is specified we use the new `remixEntryJsString` utility to write out a JS file containing the same "hydration" JS that `<Scripts>` would normally generate.